packages/taiko-client/prover/proof_submitter/proof_submitter.go

```diff
@@ -299,6 +334,145 @@ func (s *ProofSubmitter) SubmitProof(
 	return nil
 }
 
+// BatchSubmitProofs implements the Submitter interface to submit proof aggregation.
+func (s *ProofSubmitter) BatchSubmitProofs(ctx context.Context, batchProof *proofProducer.BatchProofs) error {
+	log.Info(
+		"Batch submit block proofs",
+		"proof", common.Bytes2Hex(batchProof.BatchProof),
+		"blockIds", batchProof.BlockIDs,
+		"tier", batchProof.Tier,
+	)
+	var (
+		invalidProofs       []*proofProducer.ProofWithHeader
+		latestProvenBlockID = big.NewInt(0)
+	)
+	if len(batchProof.Proofs) == 0 {
+		return proofProducer.ErrInvalidLength
+	}
+	// Check if the proof has already been submitted.
+	proofStatus, err := rpc.BatchGetBlocksProofStatus(
+		ctx,
+		s.rpc,
+		batchProof.BlockIDs,
+		batchProof.Proofs[0].Opts.ProverAddress,
+		s.proverSetAddress,
+	)
+	if err != nil {
+		return err
+	}
+	stateVars, err := s.rpc.GetProtocolStateVariables(&bind.CallOpts{Context: ctx})
+	if err != nil {
+		log.Warn(
+			"Failed to fetch state variables",
+			"error", err,
+		)
+		return err
+	}
+	for i, proof := range batchProof.Proofs {
+		// Check if this proof is still needed to be submitted.
+		ok, err := s.sender.ValidateProof(ctx, proof, new(big.Int).SetUint64(stateVars.B.LastVerifiedBlockId))
+		if err != nil {
+			return err
+		}
+		if !ok {
+			log.Error("a valid proof for block is already submitted", "blockId", proof.BlockID)
+			invalidProofs = append(invalidProofs, proof)
+			continue
+		}
+
+		if proofStatus[i].IsSubmitted && !proofStatus[i].Invalid {
+			log.Error("a valid proof for block is already submitted", "blockId", proof.BlockID)
+			invalidProofs = append(invalidProofs, proof)
+			continue
+		}
+
+		// Get the corresponding L2 block.
+		block, err := s.rpc.L2.BlockByHash(ctx, proof.Header.Hash())
+		if err != nil {
+			log.Error("failed to get L2 block with given hash",
+				"hash", proof.Header.Hash(),
+				"error", err,
+			)
+			invalidProofs = append(invalidProofs, proof)
+			continue
+		}
+
+		if block.Transactions().Len() == 0 {
+			log.Error("invalid block without anchor transaction, blockID", "blockId", proof.BlockID)
+			invalidProofs = append(invalidProofs, proof)
+			continue
+		}
+
+		// Validate TaikoL2.anchor transaction inside the L2 block.
+		anchorTx := block.Transactions()[0]
+		if err = s.anchorValidator.ValidateAnchorTx(anchorTx); err != nil {
+			log.Error("invalid anchor transaction", "error", err)
+			invalidProofs = append(invalidProofs, proof)
+		}
+		if proof.BlockID.Cmp(latestProvenBlockID) > 0 {
+			latestProvenBlockID = proof.BlockID
+		}
+	}
+
+	if len(invalidProofs) > 0 {
+		s.proofBuffer.ClearItems(invalidProofs...)
+		return ErrInvalidProof
+	}
+
+	// Build the TaikoL1.proveBlocks transaction and send it to the L1 node.
+	if err := s.sender.SendBatchProof(
+		ctx,
+		s.txBuilder.BuildProveBlocks(batchProof, s.graffiti),
+		batchProof,
```

Comment:
Seems here `batchProof` may also still have some proved block's proof right? is it ok?